File: src/com/suushiemaniac/lang/json/value/JsonArray.java
package com.suushiemaniac.lang.json.value;

import com.suushiemaniac.lang.json.JSON;
import com.suushiemaniac.lang.json.exception.JsonValueTypeException;
import com.suushiemaniac.lang.json.util.StringUtils;
import com.suushiemaniac.lang.json.exception.JsonNotIterableException;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class JsonArray extends JsonElement {
    private List<JSON> elements;

    public JsonArray(List<JSON> elements) {
        this(null, elements);
    }

    public JsonArray(JSON parent, List<JSON> elements) {
        super(parent);

        for (JSON jsonType : elements) jsonType.setParent(this);

        this.elements = elements;
    }

    @Override
    public JSON get(JSON keyIndex) {
        if (keyIndex instanceof JsonNumber) return this.elements.get(keyIndex.intValue());
        else throw new JsonNotIterableException("Can't get child element by given JSON type. Expected: Number, Found: " + keyIndex.type());
    }

	@Override
	public JSON get(String key) {
		try {
			return this.elements.get(Integer.parseInt(key));
		} catch (NumberFormatException e) {
			throw new JsonValueTypeException("Can't get child element by non-numeric string key " + key);
		}
	}

	@Override
    public JSON get(int index) {
        return this.elements.get(index);
    }

    @Override
    public void set(JSON keyIndex, JSON value) {
        value.setParent(this);
        if (keyIndex instanceof JsonNumber) this.elements.set(keyIndex.intValue(), value);
    }

	@Override
	public void set(String key, JSON value) {
		try {
			this.elements.set(Integer.parseInt(key), value);
			value.setParent(this);
		} catch (NumberFormatException e) {
			throw new JsonValueTypeException("Can't set list element by non-numeric string key " + key);
		}
	}

	@Override
    public void set(int index, JSON value) {
        value.setParent(this);
        this.elements.set(index, value);
    }

    @Override
    public void set(JSON value) {
        if (value instanceof JsonArray) this.elements = ((JsonArray) value).elements;

        for (JSON t : this.elements) t.setParent(this);
    }

    @Override
    public void add(JSON value) {
        value.setParent(this);
        this.elements.add(value);
    }

	@Override
	public void add(JSON keyIndex, JSON value) {
		if (keyIndex instanceof JsonNumber) {
			this.add(keyIndex.intValue(), value);
		} else {
			throw new JsonValueTypeException("Can't add child element by given JSON type. Expected: Number, Found: " + keyIndex.type());
		}
	}

	@Override
	public void add(String key, JSON value) {
		try {
			this.elements.add(Integer.parseInt(key), value);
			value.setParent(this);
		} catch (NumberFormatException e) {
			throw new JsonValueTypeException("Can't add child element by non-numeric string key " + key);
		}
	}

	@Override
	public void add(int index, JSON value) {
		this.elements.add(index, value);
		value.setParent(this);
	}

	@Override
    public void remove(JSON value) {
        this.elements.remove(value);
    }

    @Override
    public void remove(int index) {
        this.elements.remove(index);
    }

	@Override
	public void remove(String key) {
		throw new JsonValueTypeException("Can't remove child element by string key");
	}

	@Override
    public JSON keyIndexOf(JSON content) {
        return new JsonInteger(this.elements.indexOf(content));
    }

    @Override
    public void clear() {
        this.elements.clear();
    }

    @Override
    public int size() {
        return this.elements.size();
    }

    @Override
    public int deepSize() {
        int deepSize = 0;

        for (JSON t : this.elements)
        	deepSize += t.deepSize();

        return deepSize;
    }

    @Override
    public Iterator<JSON> iterator() {
        return this.elements.iterator();
    }

    public String toString() {
        List<String> stringedList = new ArrayList<>();

        for (JSON t : this.elements)
        	stringedList.add(t.toString());

        return "[" + String.join(",", stringedList) + "]";
    }

    @Override
    public boolean equals(Object other) {
        return other instanceof JsonArray && ((JsonArray) other).elements.equals(this.elements);
    }

    //TODO maybe simplify
    @Override
    public String toFormatString() {
        List<String> stringedList = new ArrayList<>();

        for (JSON t : this.elements)
			stringedList.add(this.size() <= 1 ? StringUtils.reduceTab(t.toFormatString()) : t.toFormatString());

        String openTabbing = StringUtils.copy("\t", this.hierarchy() - (this.parent() instanceof JsonElement && this.parent().size() <= 1 ? 1 : 0));
        String borderBreak = stringedList.size() <= 1 ? "" : "\n";
        String closingTabs = stringedList.size() <= 1 ? "" : openTabbing;
        return openTabbing + "[" + borderBreak + String.join(",\n", stringedList) + borderBreak + closingTabs + "]";
    }

    @Override
    public String toXMLString() {
        List<String> stringedList = new ArrayList<>();

        String key = "array";
        JSON parent = this.parent();
        if (parent instanceof JsonObject) {
            for (JSON parentKey : parent.keySet()) {
                if (parent.get(parentKey).equals(this)) {
                    key = parentKey.stringValue();
                }
            }
        }

        for (JSON t : this.elements) {
            String openTag = "<" + key + ">";
            String closeTag = "</" + key + ">";

            String val = t.toXMLString();
            if (val.length() > 0) stringedList.add(openTag + val + closeTag);
        }

        return String.join("\n", stringedList);
    }

	@Override
	public Object toNative() {
		return this.nativeList();
	}

	@Override
    public Collection<JSON> collect() {
        return new ArrayList<>(this.elements);
    }

	@Override
    public Set<JSON> keySet() {
    	Set<JSON> keySet = new HashSet<>();

		for (int i = 0; i < this.elements.size(); i++) {
			keySet.add(new JsonInteger(i));
		}

        return keySet;
    }

	@Override
	public Map<String, Object> nativeMap() {
		Map<String, Object> nativeMap = new HashMap<>();

		for (int i = 0; i < this.elements.size(); i++) {
			nativeMap.put("" + i, this.elements.get(i).toNative());
		}

		return nativeMap;
	}

	@Override
	public List<Object> nativeList() {
		return this.elements.stream().map(JSON::toNative).collect(Collectors.toList());
	}

	@Override
	public Set<Object> nativeSet() {
		return this.elements.stream().map(JSON::toNative).collect(Collectors.toSet());
	}
}

File: src/main/kotlin/com/suushiemaniac/lang/json/value/JsonArray.kt
package com.suushiemaniac.lang.json.value

import com.suushiemaniac.lang.json.JSON
import com.suushiemaniac.lang.json.exception.JsonValueTypeException
import com.suushiemaniac.lang.json.util.StringUtils.reduceTab
import com.suushiemaniac.lang.json.exception.JsonNotIterableException

class JsonArray(parent: JSON?, private var elements: MutableList<JSON>) : JsonElement(parent) {
    constructor(elements: MutableList<JSON>) : this(null, elements)

    init {
        for (jsonType in elements) jsonType.parent = this
    }

    override fun get(keyIndex: JSON): JSON? {
        return if (keyIndex is JsonNumber)
            this.elements[keyIndex.intValue()]
        else
            throw JsonNotIterableException("Can't get child element by given JSON type. Expected: Number, Found: " + keyIndex.type())
    }

    override fun get(key: String): JSON? {
        try {
            return this.elements[Integer.parseInt(key)]
        } catch (e: NumberFormatException) {
            throw JsonValueTypeException("Can't get child element by non-numeric string key $key")
        }

    }

    override fun get(index: Int): JSON? {
        return this.elements[index]
    }

    override fun set(keyIndex: JSON, value: JSON) {
        value.parent = this
        if (keyIndex is JsonNumber) this.elements[keyIndex.intValue()] = value
    }

    override fun set(key: String, value: JSON) {
        try {
            this.elements[key.toInt()] = value
            value.parent = this
        } catch (e: NumberFormatException) {
            throw JsonValueTypeException("Can't set list element by non-numeric string key $key")
        }

    }

    override fun set(index: Int, value: JSON) {
        value.parent = this
        this.elements[index] = value
    }

    override fun set(value: JSON) {
        if (value is JsonArray) {
            this.elements = value.elements
        }

        for (t in this.elements) {
            t.parent = this
        }
    }

    override fun add(value: JSON) {
        value.parent = this
        this.elements.add(value)
    }

    override fun add(keyIndex: JSON, value: JSON) {
        if (keyIndex is JsonNumber) {
            this.add(keyIndex.intValue(), value)
        } else {
            throw JsonValueTypeException("Can't add child element by given JSON type. Expected: Number, Found: " + keyIndex.type())
        }
    }

    override fun add(key: String, value: JSON) {
        try {
            this.elements.add(key.toInt(), value)
            value.parent = this
        } catch (e: NumberFormatException) {
            throw JsonValueTypeException("Can't add child element by non-numeric string key $key")
        }

    }

    override fun add(index: Int, value: JSON) {
        value.parent = this
        this.elements.add(index, value)
    }

    override fun remove(value: JSON) {
        this.elements.remove(value)
    }

    override fun remove(index: Int) {
        this.elements.removeAt(index)
    }

    override fun remove(key: String) {
        throw JsonValueTypeException("Can't remove child element by string key")
    }

    override fun keyIndexOf(content: JSON): JSON {
        return JsonInteger(this.elements.indexOf(content))
    }

    override fun clear() {
        this.elements.clear()
    }

    override fun size(): Int {
        return this.elements.size
    }

    override fun deepSize(): Int {
        return this.elements.sumBy { it.deepSize() }
    }

    override fun iterator(): Iterator<JSON> {
        return this.elements.iterator()
    }

    override fun toString(): String {
        return this.elements.joinToString(",", prefix = "[", postfix = "]")
    }

    override fun equals(other: Any?): Boolean {
        return other is JsonArray && other.elements == this.elements
    }

    override fun hashCode(): Int {
        return this.elements.hashCode()
    }

    //TODO maybe simplify
    override fun toFormatString(): String {
        val stringedList = this.elements.map { if (this.size() <= 1 || this.depth() <= 1) it.toFormatString().reduceTab() else it.toFormatString() }

        val openTabbing = "\t".repeat(this.hierarchy() - if (this.parent is JsonElement && this.parent!!.size() <= 1) 1 else 0)

        val borderBreak = if (stringedList.size <= 1 || this.depth() <= 1) "" else "\n"
        val closingTabs = if (stringedList.size <= 1 || this.depth() <= 1) "" else openTabbing

        return openTabbing + "[" + borderBreak + stringedList.joinToString(",$borderBreak") + borderBreak + closingTabs + "]"
    }

    override fun toXMLString(): String {
        val stringedList = ArrayList<String>()

        var key = "array"
        val parent = this.parent()

        if (parent is JsonObject) {
            for (parentKey in parent.keySet()) {
                if (parent[parentKey] == this) {
                    key = parentKey.stringValue()
                }
            }
        }

        for (t in this.elements) {
            val xmlVal = t.toXMLString()
            if (xmlVal.isNotEmpty()) stringedList.add("<$key>$xmlVal</$key>")
        }

        return stringedList.joinToString("\n")
    }

    override fun toNative(): Any {
        return this.nativeList()
    }

    override fun nativeKeySet(): Set<String> {
        return this.nativeMap().keys
    }

    override fun collect(): Collection<JSON> {
        return this.elements.toList()
    }

    override fun keySet(): Set<JSON> {
        return this.elements.indices.map { JsonInteger(it) }.toSet()
    }

    override fun <T> nativeMap(mapping: (JSON) -> T): Map<String, T> {
        return this.elements.mapIndexed { idx, json -> idx.toString() to mapping(json) }.toMap()
    }

    override fun <T> nativeList(mapping: (JSON) -> T): List<T> {
        return this.elements.map(mapping).toList()
    }

    override fun <T> nativeSet(mapping: (JSON) -> T): Set<T> {
        return this.elements.map(mapping).toSet()
    }
}